pootle/apps/virtualfolder/receivers.py

```diff
@@ -19,85 +20,67 @@ def update_vfolder_tree(vf, store):
-            if not pootle_path.startswith(location):
-                continue
-
-            for filename in vf.filter_rules.split(","):
-                if pootle_path == "".join([location, filename]):
-                    vf.stores.add(store)
-                    store_added = True
-                    break
-
-            if store_added:
-                break
-
-        if store_added:
-            update_vfolder_tree(vf, store)
+@receiver(post_save, sender=Store)
+def handle_store_save(sender, instance, created, **kwargs):
+    if not created:
+        return
+    project = instance.translation_project.project
+    language = instance.translation_project.language
+
+    # limit to vfolders that could be possibly related
+    vfolders = (
+        project.vfolders.filter(all_languages=True)
+        | language.vfolders.filter(all_projects=True)
+        | project.vfolders.filter(languages=language)
+        | VirtualFolder.objects.filter(
+            all_languages=True, all_projects=True))
+    for vf in vfolders:
+        added = vf.path_matcher.add_store_if_matching(instance)
+        if added:
+            update_vfolder_tree(vf, instance)
+            instance.set_priority()
 
 
 @receiver(post_save, sender=VirtualFolder)
-def vfolder_save_handler(sender, instance, created, **kwargs):
-    """Remove Units from VirtualFolder when vfolder changes
-
-    - Check the original VirtualFolder object's locations
-    - Check the new VirtualFolders object's locations
-    - Remove any units from the VirtualFolder that are only in original's
-      locations and reset Unit priority
-    - Update VirtualFolderTree for any Stores that may have been affected
-    """
-    locations = set()
-    for location in instance.all_locations:
-        for filename in instance.filter_rules.split(","):
-            locations.add("".join([location, filename]))
-
-    stores_we_want = Store.objects.none()
-    for location in locations:
-        stores_we_want = stores_we_want | Store.objects.filter(
-            pootle_path__startswith=location)
-    to_remove = list(
-        instance.stores.exclude(
-            pk__in=stores_we_want.values_list("pk", flat=True)))
-    to_add = list(
-        stores_we_want.exclude(
-            pk__in=instance.stores.values_list("pk", flat=True)))
-    instance.stores.remove(*to_remove)
-    instance.stores.add(*to_add)
-    for store in to_add:
+def handle_vfolder_save(sender, instance, created, **kwargs):
+    added, deleted = instance.path_matcher.update_matching_stores()
+    for store in added:
         update_vfolder_tree(instance, store)
-    for store in to_remove:
-        store.set_priority()
-    for store in stores_we_want:
+        if store.priority != instance.priority:
+            store.set_priority()
+    stores = Store.objects.filter(
+        pk__in=deleted).exclude(priority=instance.priority)
+    for store in stores:
         store.set_priority()
 
 
-@receiver(post_save, sender=Store)
```

Comment:
Just saw it was moved up. This just increased the diff and mades more difficult to review.